Grading Performance Work

ormsbee opened the pull request:
This builds off of @adampalay's work in #7168

Features:
* Max scores cache to limit XModule instantiation to find out the possible score (cache is automatically invalidated when an new version of the course is pushed).
* Filtering to limit the amount of data we pull back in the `FieldDataCache` and tree crawling to only be items that might impact grading.
* Request-level caching of the partitioning service (this was causing many requests for A/B tests)
* Remove many of the StudentModule queries that weren't using the `FieldDataCache`.

On a local test of 8.MechCx (a good stress test, as it has > 800 problems and uses A/B testing more than most), this brings down the progress page render time from 60s (!) to 20s in an uncached state, and 5s in a cached state. Leaning more heavily on `FieldDataCache` means that we go from potentially thousands of CSM queries on particularly large courses, down to one or two. There's clearly more work to do, but this should help curb the grading time on some of our largest courses and significantly reduce CSM usage (about half of CSM usage happens through the progress page).

@cpennington, @adampalay: Review please?

ormsbee commented on lms/djangoapps/courseware/tests/test_submitting_problems.py:
```diff
@@ -426,6 +438,36 @@ def test_submission_show_answer(self):
         )
         self.assertEqual(json.loads(resp.content).get("success"), err_msg)
 
+    @override_settings(KEY_PREFIX=str(uuid.uuid4()))  # hack: we do this to ensure a unique cache per test
```
So after reading a bit more, I don't think this actually works. Will look over the tests again...

adampalay replied:
In this case, since we know only `location_to_cache` is being added to the cache, we can just remove it after the test.

ormsbee commented on common/lib/xmodule/xmodule/partitions/partitions_service.py:
```diff
@@ -4,7 +4,7 @@
 """
 from abc import ABCMeta, abstractproperty
-
+from request_cache.middleware import RequestCache
```
Argh. I forgot we can't have Django in here. Need to rethink this... :-(

ormsbee replied:
I'll just do the same thing it's doing now for the tracking func...

adampalay commented on lms/djangoapps/courseware/grades.py:
```diff
@@ -152,6 +223,17 @@ def _grade(student, request, course, keep_raw_scores):
+                course.id, student, course, depth=None, descriptor_filter=descriptor_affects_grading
+            )
+
+    max_scores_cache = MaxScoresCache(course.subtree_edited_on.isoformat())
```
Nice, so this way the cache resets every time the course is updated?

ormsbee replied:
Yup, that's the idea. I set the timeout for a day otherwise, just in case something gets borked. Still need to add tests around this behavior, but it seems to work.

adampalay commented on lms/djangoapps/courseware/management/commands/grade_students.py:
```diff
@@ -0,0 +1,19 @@
+from django.core.management.base import BaseCommand
```
I was just using this file for debugging, this file should be deleted

adampalay commented on lms/djangoapps/courseware/model_data.py:
```diff
@@ -64,6 +64,7 @@ def __init__(self, descriptors, course_id, user, select_for_update=False, asides
         self.cache = {}
         self.descriptors = descriptors
         self.select_for_update = select_for_update
+        self.scores = {}
```
should this more properly be something like `modules_with_scores`? Just calling it `scores` is a bit confusing

adampalay commented on lms/djangoapps/courseware/model_data.py:
```diff
@@ -77,7 +78,12 @@ def __init__(self, descriptors, course_id, user, select_for_update=False, asides
+                    self.cache[cache_key] = field_object
+                    if scope == Scope.user_state:
+                        student_module = field_object
+                        scope, location = cache_key
```
nit: `__, location = cache_key`

adampalay commented on lms/djangoapps/courseware/tests/test_grades.py:
```diff
@@ -121,3 +127,80 @@ def _gradesets_and_errors_for(self, course_id, students):
+        self.request = RequestFactory().get('/')
+        self.locations = [problem.location for problem in self.problems]
+
+    @override_settings(KEY_PREFIX=str(uuid.uuid4()))  # hack to ensure unique cache per test
```
instead of doing this, we can just  explicitly  remove any items we added to the cache after the test runs. Not a great solution, but better than overriding the `KEY_PREFIX`.

ormsbee replied:
The `MaxScoresCache` itself takes a prefix to do the course publish timestamp expiration thing, so I can just use that and not worry about collisions. Will get back to this tonight.

ormsbee commented on common/lib/xmodule/xmodule/partitions/partitions_service.py:
```diff
@@ -55,7 +60,10 @@ def get_user_group_id_for_partition(self, user_partition_id):
+        self._request_cache[cache_key] = group_id
+
+        return group_id
 
```
@jimabramson: Tagging you here on the PartitionService caching.

ormsbee commented on common/lib/xmodule/xmodule/partitions/tests/test_partitions.py:
```diff
@@ -325,9 +325,19 @@ def test_get_user_group_id_for_partition(self):
+            course_id=self.course.id,
+            track_function=Mock()
+        )
+        group2_id = partition_service2.get_user_group_id_for_partition(user_partition_id)
```
So this is something I need to clean up and better message in the tests. The reason this is here is because of the cache on a `PartitionService`. If it doesn't get a cache object explicitly passed in, it will create one of its own, so it will last the lifetime of that object. In the actual runtime, we're passing in a request-cache so that every instance of PartitionService that's made (and there can be a lot of them) share that cached info -- but the cache dies at the end of the request.

ormsbee replied:
But that's ok in terms of system behavior, right? Does the code rely on the fact that you can change a user's partitioning scheme and read it back in the same request?

jimabramson replied:
PartitionSchemes (at least, "real" ones) are read-only adapter interfaces to something else.  That is, there's nothing saying that data used by a specific partition scheme couldn't change during any request or other arbitrary lifecycle - the partition scheme doesn't make any assumptions about it, and the partition service just forwards what the scheme says.

Which is why I think a better implementation, if possible, would be to disable caching unless a request cache is passed in.  The passing in of a cache container would indicate that the caller intends to consider the underlying data static/immutable throughout the lifetime of the service, which seems like the behavior you might want during a grading run.  Does that make sense?

ormsbee replied:
@jimabramson: ^

jimabramson replied:
@ormsbee ?

ormsbee replied:
Sorry, I hadn't seen your message when I wrote that.

So I can definitely make the default behavior non-caching unless an explicit cache object is passed in. This thing is only created once in our app, in the `LmsModuleSystem` constructor. I still have the same question, though perhaps pushed one level up the stack. It looks like the pieces that are doing the real work (`get_group_for_user` in `RandomUserPartitionScheme` and `CohortPartitionScheme`) will both set a value if there exists None beforehand. I'm assuming that the values they spit back within a given request are going to stay the same. I *think* that's safe, but I'm really not familiar with this part of the code.

jimabramson replied:
i thought at least in the case of CohortPartitionScheme that there were parameters to cancel the auto-assignment side effect in get group for user - maybe i'm misremembering?  Either way, in practice, i do think that within the lifecycle of a web request the possibility of a return value from either of the implemented PartitionSchemes changing is very close to zero.  It's only during other as-yet-unimagined, longer-running transactions or tasks where i think mandatory caching could become confusing or limiting.

cpennington commented on lms/djangoapps/courseware/grades.py:
```diff
@@ -26,10 +27,97 @@
+        return cls(course.subtree_edited_on.isoformat())
+
+    def fetch_from_remote(self, locations):
+        """
```
Is this something that you're always going to call right after `__init__`?

ormsbee replied:
I don't know honestly. I thought that having it separate could make testing easier and be a parallel to the pushing code, but maybe that's silly.

cpennington commented on lms/djangoapps/courseware/grades.py:
```diff
@@ -255,11 +350,14 @@ def create_module(descriptor):
     letter_grade = grade_for_percentage(course.grade_cutoffs, grade_summary['percent'])
     grade_summary['grade'] = letter_grade
-    grade_summary['totaled_scores'] = totaled_scores  	# make this available, eg for instructor download & debugging
+    grade_summary['totaled_scores'] = totaled_scores    # make this available, eg for instructor download & debugging
```
Nit: unnecessary whitespace change?

ormsbee replied:
Not really sure how that got there, but I'll change it back.

cpennington commented on lms/djangoapps/courseware/grades.py:
```diff
@@ -255,11 +350,14 @@ def create_module(descriptor):
         # so grader can be double-checked
         grade_summary['raw_scores'] = raw_scores
+
+    max_scores_cache.push_to_remote()
```
Any time I see a pattern like this (setup, then call some cleanup later), it makes me wonder if it should be a context manager.

ormsbee replied:
Yeah, that occurred to me, but the indentation on the grading code is already 5-6 levels deep, and I didn't want to add a level.

ormsbee replied:
And I was too lazy to refactor further :P

cpennington commented on lms/djangoapps/courseware/model_data.py:
```diff
@@ -77,7 +78,11 @@ def __init__(self, descriptors, course_id, user, select_for_update=False, asides
+                    cache_key = self._cache_key_from_field_object(scope, field_object)
+                    self.cache[cache_key] = field_object
+                    if scope == Scope.user_state:
+                        _scope, location = cache_key
```
This could just be: `location = field_object.module_state_key`, right?

ormsbee replied:
Right. I originally did it this way because I incorrectly thought `field_object.module_state_key` would come back as a unicode object. Changing.

cpennington commented:
Some small notes, but no major concerns.

ormsbee commented:
@cpennington, @adampalay: I think I've responded to your comments. I didn't do the context manager thing mostly for extra indent. Is that something you want to see before it goes in?

Assuming devops can get the prod replica sorted, I'm going to try some larger scale testing over the weekend. Could you please give :+1: if it's ok to go?

cpennington commented:
No, the contextmanager isn't a requirement for merge. I can definitely understand not wanting more nesting

adampalay commented on lms/djangoapps/courseware/tests/test_grades.py:
```diff
@@ -3,13 +3,16 @@
 """
 from django.http import Http404
 from django.test.utils import override_settings
+from django.test.client import RequestFactory
```
Is this an unused import?

adampalay commented:
:+1:

ormsbee commented:
This gives about a 3X speed up on load-tests against the demo course. It also gives a lot more regularity: median rendering time went from 750ms to 240ms, but 99th percentile times went from 1400ms to 270ms.

Still running a batch of existing courses through grading runs with old and new to verify that we're producing the same results. If so, this should be able to go out in next week's release cycle.

adampalay commented:
That's fantastic!!!!

ormsbee commented:
<sigh> There is something about this PR that results in slight course-tree differences for some students in one of the test courses. @nasthagiri: I'm at a conference for the next couple of days. I'm going to try to track this problem down tomorrow night and fix it, but even if I do so, it'll take another day to run the big regression test again. It's unlikely that I'll get to merge this until Friday.

nasthagiri commented:
@ormsbee No problem.  In that case, I have separated out the partition cache changes into its own commit and cherry-picked it into my PR: https://github.com/edx/edx-platform/pull/7298/commits
So whoever merges first can commit those changes.

ormsbee commented:
Okay, latest update -- I thought it was extra columns because of the length discrepancy, but it was really because we were assigning the score = 0 when it should have been 0.0 (int vs. float). Going to make that change and kick things off on the sandbox again.

ormsbee commented on lms/djangoapps/courseware/grades.py:
```diff
@@ -427,18 +559,24 @@ def get_score(course_id, user, problem_descriptor, module_creator, scores_cache=
         total = student_module.max_grade
+
+    elif max_score is not None and settings.FEATURES.get("ENABLE_MAX_SCORE_CACHE"):
+        correct = 0
```
Should be 0.0 for consistency. Changing...

ormsbee commented on lms/djangoapps/courseware/grades.py:
```diff
@@ -427,18 +559,24 @@ def get_score(course_id, user, problem_descriptor, module_creator, scores_cache=
+    # seen an older version of the problem -- they're still graded on what was
+    # possible when they tried the problem, not what it's worth now.
     if student_module is not None and student_module.max_grade is not None:
         correct = student_module.grade if student_module.grade is not None else 0
```
This should be 0.0 too... even though that alters existing behavior. <sigh>

ormsbee commented:
@dan-f

ormsbee commented:
test this

ormsbee commented:
test this please

ormsbee commented:
test this please

mattdrayer commented:
Tagging myself on this PR so I can find it easily:  @mattdrayer

ormsbee commented:
okay, going to rebase this one more time...

cpennington commented on common/lib/xmodule/xmodule/course_module.py:
```diff
@@ -1316,11 +1317,21 @@ def grading_context(self):
+            cat for (cat, xblock_class) in XBlock.load_classes() if (
+                getattr(xblock_class, 'has_score', True) or getattr(xblock_class, 'has_children', True)
+            )
+        )
```
Could this just be a constant?

ormsbee replied:
You mean at the module level? I guess I don't really understand how extensions are loaded under the covers -- will `load_classes()` return the right thing if I ran this code at the time of module import?

ormsbee replied:
I moved this to be a lazy attribute on the `CourseDescriptor`. Trying to set it at the module level caused import errors.

cpennington commented on lms/djangoapps/courseware/grades.py:
```diff
@@ -26,10 +27,123 @@
+    since those children might have scores. We can avoid things like Videos,
+    which have state but cannot ever impact someone's grade.
+    """
+    return getattr(descriptor, "has_score", False) or getattr(descriptor, "has_children", False)
```
This is the same code as in `course_module.py`.

ormsbee commented:
test this please

ormsbee commented:
@cpennington: I fixed the redundant code issue you pointed out. I've rebased, tests pass once again, and if this thing passes and I don't get objections, I plan to hit the merge button tomorrow morning...

adampalay commented on lms/djangoapps/courseware/grades.py:
```diff
@@ -9,28 +10,149 @@
 from xmodule.graders import Score
 from xmodule.modulestore.django import modulestore
 from xmodule.modulestore.exceptions import ItemNotFoundError
+#from xmodule.course_module import BLOCK_TYPES_AFFECTING_GRADING
```
delete this line

adampalay commented:
awesome

adampalay commented:
:fireworks:

mattdrayer commented:
Nice!

nasthagiri commented:
Yippee!
